### DISCUSSION TITLE
Is expo bare workflow the same as a bare project that is created by react-native-cli?

### DISCUSSION THREAD

**Discussion by mattchannn** [expo/expo#17689] [Archived: Questions] [answered]
Hi all,

As far as I am concerned, there are 4 ways to create a React Native app

1. expo managed workflow
2. expo bare workflow
3. react native cli
4. react native boilerplate

I think the question should be is expo bare workflow 100% the same as a bare project that is created by react native cli or copied from a react native boilerplate? Can they be treated as a bare project, the same? I would like a true answer since I have seen many debates over Stackoverflow and Reddit. Some says "expo bare workflow vs react native cli" while some says "when you eject from expo managed workflow, you can still have a bare react native app". I don't know which one is correct..

**Comment by barthap** [comment, marked as answer]
> **Update:** See [Brent's comment](https://github.com/expo/expo/discussions/17689#discussioncomment-2949433) below

I would say that _bare workflow_ is when you have a _bare React Native_ app (created by react-native-cli or in any other way) So
```sh
# create a bare RN app using react-native-cli
npx react-native init MyApp && cd MyApp

# add expo, afaik there's also a command `npx install-expo-modules`
yarn add expo

# now you still have a bare project, but you can use expo libraries now
npx expo install expo-media-library
```
or do the same using `expo-cli`:
```sh
expo init MyApp -t expo-template-bare-minimum && cd MyApp
npx expo install expo-media-library
```

> Some says "expo bare workflow vs react native cli" while some says "when you eject from expo managed workflow, you can still have a bare react native app". I don't know which one is correct..

- In bare project it doesn't matter if you use react-native cli or expo-cli
- _"when you eject from expo managed workflow, you can still have a bare react native app"_ - true.

**Reply by mattchannn** [reply]
@barthap So could we say such statement / comparison `expo bare workflow vs react native cli` in Stackoverflow is incorrect as these refers to the same thing.

**Reply by barthap** [reply]
Generally yes, these terms are not strictly defined and can sometimes be confusing.
"Choosing between RN CLI and Expo" is a very harmful myth, because Expo is a wide term and consists of many elements (expo go, expo-cli, SDK, dev-client, EAS) - you can mix and match them in many ways. A year ago I wrote a blog post about it: https://blog.swmansion.com/top-7-myths-about-expo-in-2021-ee66f929bf9f

**Reply by barthap** [reply]
So you can for example:
- have a bare RN app (created using react-native-cli) and then switch to expo-cli (and mostly vice versa)
- use react-native-cli (or expo-cli) with expo custom development build (your personal version of expo go, see expo-dev-client)
- use react-native-cli (or expo-cli) with any Expo SDK libraries
- create a expo managed app (`expo init`) and then do `expo prebuild` to generate a bare app (create `ios` and `android` directories). Then you can delete them and do it again as many times as you wish. You can still use expo-cli or switch to react-native-cli.

**Reply by brentvatne** [reply]
To summarize and clarify the discussion here, for our purposes we can think of there as being two types of React Native projects:

1. **Developer only writes JS**: the native projects (`ios` and `android` directories) are generated on demand when building and the developer only writes JavaScript code and makes modifications to native projects through config plugins if needed. While there may at some point be other ways to accomplish this, the only way that exists right now is the Expo managed workflow.
2. **Developer writes JS and native code**: the native projects (`ios` and `android` directories) are modified manually by the developer and are committed to source control. This is what we call the bare workflow, in order to differentiate it from the managed workflow.

Whether you have Expo modules installed in my opinion is not relevant to the bare workflow. I say this because if you install Expo modules in a project created by React Native Community CLI you're not meaningfully changing the way you work, you're just introducing some new modules into your app. So, I think you have made an astute observation @mattchannn that whether you initialize a project with Expo CLI with bare workflow or React Native Community CLI they are fundamentally the same thing. The only difference is whether the `expo` package is installed yet.

**Reply by mattchannn** [reply]
@brentvatne @barthap Thank you for your answer! Yes I think these answers clarify my understanding to different types of React Native project. I will mark this as an accepted answer and hopefully others could gain insight from this discussion thread. Cheers!

**Comment by Jscripter-pk** [comment]
@mattchannn As after getting answer on this did you try eas update with you react-native cli created app? If yes can you please guide a bit on how you go and implement things for you.
As I have project that is created with react-native-init and is already deployed to their respective store and we wanted to implement OTA with EAS update.

**Reply by mattchannn** [reply]
Hi @Jscripter-pk , I never tried EAS build on my own because react native cli i.e.: bare workflow satisfied our needs already.

**Reply by Jscripter-pk** [reply]
@mattchannn yeah I am asking how you go for bare workflow like just following docs help you as after going through docs I get confused as there are too many in page linkings

**Reply by mattchannn** [reply]
@Jscripter-pk 

I personally just treat expo as a topping on ice-creaming because to me I just need the expo modules but disclaimer: I am sure Expo can do a lot more than a topping. Expo just doesn't ONLY support React Native but can encapsulate many complexities from React Native and make developers life better, of course, with pros and cons.

To go for a bare workflow, I think you can follow this [page](https://reactnative.dev/docs/environment-setup?guide=native&os=macos&platform=ios) to create your bare workflow app. Or just like mentioned above.

> create a expo managed app (expo init) and then do expo prebuild to generate a bare app (create ios and android directories)



Please also refer to this [blog post](https://blog.swmansion.com/top-7-myths-about-expo-in-2021-ee66f929bf9f) too, very detailed but I believe expo eject is deprecated now.
